File: ios/HybridNearbyConnections.swift
import NitroModules

public class HybridNearbyConnections: HybridNearbyConnectionsSpec_base, HybridNearbyConnectionsSpec_protocol {

    private var multipeerModule = MultipeerConnectivityModule()

    // MARK: - Event callbacks

    public var onPeerFound: ((_ peerId: String, _ name: String) -> Void)?
    public var onPeerLost: ((_ peerId: String) -> Void)?
    public var onInvitationReceived: ((_ peerId: String, _ name: String) -> Void)?
    public var onConnected: ((_ peerId: String, _ name: String) -> Void)?
    public var onDisconnected: ((_ peerId: String) -> Void)?
    public var onTextReceived: ((_ peerId: String, _ text: String) -> Void)?

    public override init() {
        super.init()
        multipeerModule.delegate = self
    }

    // MARK: - Methods

    public func isPlayServicesAvailable() throws -> Promise<Bool> {
        return Promise.resolved(withResult: true)
    }

    public func startAdvertise(name: String, strategy: Strategy?) throws -> Promise<String> {
        // strategy is ignored on iOS — MultipeerConnectivity has no equivalent concept
        let peerId = multipeerModule.startAdvertise(name)
        return Promise.resolved(withResult: peerId)
    }

    public func stopAdvertise() throws -> Promise<Void> {
        multipeerModule.stopAdvertise()
        return Promise.resolved()
    }

    public func startDiscovery(name: String, strategy: Strategy?) throws -> Promise<String> {
        // strategy is ignored on iOS — MultipeerConnectivity has no equivalent concept
        let peerId = multipeerModule.startDiscovery(name)
        return Promise.resolved(withResult: peerId)
    }

    public func stopDiscovery() throws -> Promise<Void> {
        multipeerModule.stopDiscovery()
        return Promise.resolved()
    }

    public func requestConnection(advertisePeerId: String) throws -> Promise<Void> {
        try multipeerModule.requestConnection(to: advertisePeerId)
        return Promise.resolved()
    }

    public func acceptConnection(targetPeerId: String) throws -> Promise<Void> {
        try multipeerModule.acceptConnection(to: targetPeerId)
        return Promise.resolved()
    }

    public func rejectConnection(targetPeerId: String) throws -> Promise<Void> {
        try multipeerModule.rejectConnection(to: targetPeerId)
        return Promise.resolved()
    }

    public func disconnect(targetPeerId: String?) throws -> Promise<Void> {
        // targetPeerId is ignored on iOS — MultipeerConnectivity disconnects the entire session
        multipeerModule.disconnect()
        return Promise.resolved()
    }

    public func sendText(targetPeerId: String, text: String) throws -> Promise<Void> {
        try multipeerModule.sendText(to: targetPeerId, payload: text)
        return Promise.resolved()
    }
}

// MARK: - NearbyConnectionCallbackDelegate

extension HybridNearbyConnections: NearbyConnectionCallbackDelegate {
    func onPeerFound(fromPeerId peerId: String, fromPeerName name: String) {
        self.onPeerFound?(peerId, name)
    }

    func onPeerLost(fromPeerId peerId: String) {
        self.onPeerLost?(peerId)
    }

    func onInvitationReceived(fromPeerId peerId: String, fromPeerName name: String) {
        self.onInvitationReceived?(peerId, name)
    }

    func onConnected(fromPeerId peerId: String, fromPeerName name: String) {
        self.onConnected?(peerId, name)
    }

    func onDisconnected(fromPeerId peerId: String) {
        self.onDisconnected?(peerId)
    }

    func onTextReceived(fromPeerId peerId: String, payload text: String) {
        self.onTextReceived?(peerId, text)
    }
}

File: android/src/main/java/com/margelo/nitro/expo/modules/nearbyconnections/HybridNearbyConnections.kt
package com.margelo.nitro.expo.modules.nearbyconnections

import android.content.Context
import android.util.Log
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.nearby.Nearby
import com.google.android.gms.nearby.connection.AdvertisingOptions
import com.google.android.gms.nearby.connection.ConnectionInfo
import com.google.android.gms.nearby.connection.ConnectionLifecycleCallback
import com.google.android.gms.nearby.connection.ConnectionResolution
import com.google.android.gms.nearby.connection.ConnectionsClient
import com.google.android.gms.nearby.connection.DiscoveredEndpointInfo
import com.google.android.gms.nearby.connection.DiscoveryOptions
import com.google.android.gms.nearby.connection.EndpointDiscoveryCallback
import com.google.android.gms.nearby.connection.Payload
import com.google.android.gms.nearby.connection.PayloadCallback
import com.google.android.gms.nearby.connection.PayloadTransferUpdate
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import expo.modules.nearbyconnections.getStrategy
import java.util.concurrent.ConcurrentHashMap

@DoNotStrip
@Keep
class HybridNearbyConnections : HybridNearbyConnectionsSpec() {

    // region Event callbacks

    override var onPeerFound: ((peerId: String, name: String) -> Unit)? = null
    override var onPeerLost: ((peerId: String) -> Unit)? = null
    override var onInvitationReceived: ((peerId: String, name: String) -> Unit)? = null
    override var onConnected: ((peerId: String, name: String) -> Unit)? = null
    override var onDisconnected: ((peerId: String) -> Unit)? = null
    override var onTextReceived: ((peerId: String, text: String) -> Unit)? = null

    // endregion

    // region State

    private var myPeerName: String = ""

    // ConcurrentHashMap — accessed from both JS thread and Nearby callback threads
    private val initiatedPeers: MutableMap<String, String> = ConcurrentHashMap()

    // endregion

    // region Nearby Connections client

    // Requires NitroModules.applicationContext to be initialized (set during React Native startup)
    private val connectionsClient: ConnectionsClient by lazy {
        Nearby.getConnectionsClient(context)
    }

    private val context: Context
        get() = requireNotNull(NitroModules.applicationContext) {
            "React Application Context is null"
        }

    // endregion

    // region Methods

    override fun isPlayServicesAvailable(): Promise<Boolean> {
        val result = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context)
        return Promise.resolved(result == ConnectionResult.SUCCESS)
    }

    override fun startAdvertise(name: String, strategy: Strategy?): Promise<String> {
        myPeerName = name
        val serviceId = context.packageName
        val options = AdvertisingOptions.Builder()
            .setStrategy(getStrategy(strategy))
            .build()

        val promise = Promise<String>()
        connectionsClient.startAdvertising(name, serviceId, advertiseCallback, options)
            .addOnSuccessListener { promise.resolve(name) }
            .addOnFailureListener { promise.reject(it) }
        return promise
    }

    override fun stopAdvertise(): Promise<Unit> {
        connectionsClient.stopAdvertising()
        return Promise.resolved(Unit)
    }

    override fun startDiscovery(name: String, strategy: Strategy?): Promise<String> {
        myPeerName = name
        val serviceId = context.packageName
        val options = DiscoveryOptions.Builder()
            .setStrategy(getStrategy(strategy))
            .build()

        val promise = Promise<String>()
        connectionsClient.startDiscovery(serviceId, discoveryCallback, options)
            .addOnSuccessListener { promise.resolve(name) }
            .addOnFailureListener { promise.reject(it) }
        return promise
    }

    override fun stopDiscovery(): Promise<Unit> {
        connectionsClient.stopDiscovery()
        return Promise.resolved(Unit)
    }

    override fun requestConnection(advertisePeerId: String): Promise<Unit> {
        val promise = Promise<Unit>()
        connectionsClient.requestConnection(myPeerName, advertisePeerId, requestConnectionCallback)
            .addOnSuccessListener { promise.resolve(Unit) }
            .addOnFailureListener { promise.reject(it) }
        return promise
    }

    override fun acceptConnection(targetPeerId: String): Promise<Unit> {
        val promise = Promise<Unit>()
        connectionsClient.acceptConnection(targetPeerId, payloadCallback)
            .addOnSuccessListener { promise.resolve(Unit) }
            .addOnFailureListener { promise.reject(it) }
        return promise
    }

    override fun rejectConnection(targetPeerId: String): Promise<Unit> {
        val promise = Promise<Unit>()
        connectionsClient.rejectConnection(targetPeerId)
            .addOnSuccessListener { promise.resolve(Unit) }
            .addOnFailureListener { promise.reject(it) }
        return promise
    }

    override fun disconnect(targetPeerId: String?): Promise<Unit> {
        if (targetPeerId != null) {
            connectionsClient.disconnectFromEndpoint(targetPeerId)
        } else {
            connectionsClient.stopAllEndpoints()
        }
        return Promise.resolved(Unit)
    }

    override fun sendText(targetPeerId: String, text: String): Promise<Unit> {
        val promise = Promise<Unit>()
        connectionsClient.sendPayload(targetPeerId, Payload.fromBytes(text.toByteArray()))
            .addOnSuccessListener { promise.resolve(Unit) }
            .addOnFailureListener { promise.reject(it) }
        return promise
    }

    // endregion

    // region Nearby Connections callbacks

    private val advertiseCallback: ConnectionLifecycleCallback =
        object : ConnectionLifecycleCallback() {
            override fun onConnectionResult(peerId: String, result: ConnectionResolution) {
                if (!result.status.isSuccess) {
                    Log.e(TAG, "Advertise connection failed for $peerId: ${result.status}")
                    return
                }
                val targetPeerName = initiatedPeers[peerId] ?: ""
                onConnected?.invoke(peerId, targetPeerName)
            }

            override fun onDisconnected(peerId: String) {
                this@HybridNearbyConnections.onDisconnected?.invoke(peerId)
            }

            override fun onConnectionInitiated(peerId: String, connectionInfo: ConnectionInfo) {
                val peerName = connectionInfo.endpointName
                initiatedPeers[peerId] = peerName
                onInvitationReceived?.invoke(peerId, peerName)
            }
        }

    private val requestConnectionCallback: ConnectionLifecycleCallback =
        object : ConnectionLifecycleCallback() {
            override fun onConnectionResult(peerId: String, result: ConnectionResolution) {
                if (!result.status.isSuccess) {
                    Log.e(TAG, "Request connection failed for $peerId: ${result.status}")
                    return
                }
                val targetPeerName = initiatedPeers[peerId] ?: ""
                onConnected?.invoke(peerId, targetPeerName)
            }

            override fun onDisconnected(peerId: String) {
                this@HybridNearbyConnections.onDisconnected?.invoke(peerId)
            }

            override fun onConnectionInitiated(peerId: String, connectionInfo: ConnectionInfo) {
                val peerName = connectionInfo.endpointName
                initiatedPeers[peerId] = peerName
                acceptConnection(peerId).catch { e ->
                    Log.e(TAG, "Auto-accept failed for $peerId", e)
                }
            }
        }

    private val discoveryCallback: EndpointDiscoveryCallback =
        object : EndpointDiscoveryCallback() {
            override fun onEndpointFound(peerId: String, info: DiscoveredEndpointInfo) {
                onPeerFound?.invoke(peerId, info.endpointName)
            }

            override fun onEndpointLost(peerId: String) {
                onPeerLost?.invoke(peerId)
            }
        }

    private val payloadCallback: PayloadCallback = object : PayloadCallback() {
        override fun onPayloadReceived(peerId: String, payload: Payload) {
            if (payload.type != Payload.Type.BYTES) return
            val bytes = payload.asBytes() ?: return
            onTextReceived?.invoke(peerId, String(bytes))
        }

        override fun onPayloadTransferUpdate(peerId: String, update: PayloadTransferUpdate) {
            // TODO: handle payload transfer update
        }
    }

    // endregion

    companion object {
        private const val TAG = "HybridNearbyConnections"
    }
}
